eap_backend/eap_api/views.py

```diff
@@ -22,225 +24,120 @@
+            return JsonResponse(serializer.data, status=201)
         return JsonResponse(serializer.errors, status=400)
 
-    # Recurse into child types.
-    name = serializer.data["name"]
-    id = serializer.data["id"]
-    success_http_code = 201
-    child_types = TYPE_DICT[obj_type]["children"]
-    for child_type in child_types:
-        if child_type not in data:
-            continue
-        for child_data in data[child_type]:
-            retval = save_json_tree(
-                child_data, child_type, parent_id=id, parent_type=obj_type
-            )
-            # If one of the subcalls returns an error, return.
-            if retval.status_code != success_http_code:
-                return retval
 
-    summary = {"name": name, "id": id}
-    return JsonResponse(summary, status=success_http_code)
+@csrf_exempt
+def user_detail(request, pk):
+    """
+    Retrieve, update, or delete a User by primary key
+    """
+    try:
+        user = EAPUser.objects.get(pk=pk)
+    except EAPUser.DoesNotExist:
+        return HttpResponse(status=404)
+    if request.method == "GET":
+        serializer = EAPUserSerializer(user)
+        user_data = serializer.data
+        return JsonResponse(user_data)
+    elif request.method == "PUT":
+        data = JSONParser().parse(request)
+        serializer = EAPUserSerializer(user, data=data, partial=True)
+        if serializer.is_valid():
+            serializer.save()
+            return JsonResponse(serializer.data)
+        return JsonResponse(serializer.errors, status=400)
+    elif request.method == "DELETE":
+        user.delete()
+        return HttpResponse(status=204)
 
 
 @csrf_exempt
+@api_view(["GET", "POST"])
+def group_list(request):
+    """
+    List all group, or make a new group
+    """
+
+    if request.method == "GET":
+        groups = get_allowed_groups(request.user)
+        serializer = EAPGroupSerializer(groups, many=True)
+        return JsonResponse(serializer.data, safe=False)
+    elif request.method == "POST":
+        data = JSONParser().parse(request)
+        data["owner_id"] = request.user.id
+        data["members"] = [request.user.id]
+        serializer = EAPGroupSerializer(data=data)
+        if serializer.is_valid():
+            serializer.save()
+            return JsonResponse(serializer.data, status=201)
+        return JsonResponse(serializer.errors, status=400)
+
+
+@csrf_exempt
+@api_view(["GET", "PUT", "DELETE"])
+def group_detail(request, pk):
+    """
+    Retrieve, update, or delete a Group by primary key
+    """
+    try:
+        group = EAPGroup.objects.get(pk=pk)
+    except EAPGroup.DoesNotExist:
+        return HttpResponse(status=404)
+    if not can_view_group(group, request.user):
+        return HttpResponse(status=403)
+    if request.method == "GET":
+        serializer = EAPGroupSerializer(group)
+        group_data = serializer.data
+        return JsonResponse(group_data)
+    elif request.method == "PUT":
```

Comment:
Are PUT and DELETE restricted to owner, or to anyone who can view?

Reply:
these should be restricted to owner, as they should be the only one for whom `can_view_group` is True.